--- net/minecraft/tileentity/TileEntityPiston.java
+++ net/minecraft/tileentity/TileEntityPiston.java
@@ -5,6 +5,8 @@
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+
+import cheeseum.pistoneverything.PistonEverything;
 import net.minecraft.block.Block;
 import net.minecraft.entity.Entity;
 import net.minecraft.nbt.NBTTagCompound;
@@ -15,6 +17,7 @@
 {
     private int storedBlockID;
     private int storedMetadata;
+    private NBTTagCompound storedTileEntityData;
 
     /** the side the front of the piston is on */
     private int storedOrientation;
@@ -39,6 +42,11 @@
         this.shouldHeadBeRendered = par5;
     }
 
+    public void storeTileEntity(NBTTagCompound tileEntityData)
+    {
+    	this.storedTileEntityData = tileEntityData;
+    }
+    
     public int getStoredBlockID()
     {
         return this.storedBlockID;
@@ -152,6 +160,11 @@
 
             if (this.worldObj.getBlockId(this.xCoord, this.yCoord, this.zCoord) == Block.pistonMoving.blockID)
             {
+            	if (this.storedTileEntityData != null) {
+                	PistonEverything.restoreStoredPistonBlock(
+                			this.worldObj, this.xCoord, this.yCoord, this.zCoord,
+                			this.storedBlockID, this.storedMetadata, this.storedTileEntityData);
+                }
                 this.worldObj.setBlock(this.xCoord, this.yCoord, this.zCoord, this.storedBlockID, this.storedMetadata, 3);
                 this.worldObj.notifyBlockOfNeighborChange(this.xCoord, this.yCoord, this.zCoord, this.storedBlockID);
             }
@@ -174,6 +187,12 @@
 
             if (this.worldObj.getBlockId(this.xCoord, this.yCoord, this.zCoord) == Block.pistonMoving.blockID)
             {
+                if (this.storedTileEntityData != null) {
+                	PistonEverything.restoreStoredPistonBlock(
+                			this.worldObj, this.xCoord, this.yCoord, this.zCoord,
+                			this.storedBlockID, this.storedMetadata, this.storedTileEntityData);
+                }
+                
                 this.worldObj.setBlock(this.xCoord, this.yCoord, this.zCoord, this.storedBlockID, this.storedMetadata, 3);
                 this.worldObj.notifyBlockOfNeighborChange(this.xCoord, this.yCoord, this.zCoord, this.storedBlockID);
             }
@@ -205,6 +224,10 @@
         this.storedOrientation = par1NBTTagCompound.getInteger("facing");
         this.lastProgress = this.progress = par1NBTTagCompound.getFloat("progress");
         this.extending = par1NBTTagCompound.getBoolean("extending");
+        if (par1NBTTagCompound.hasKey("blockTileEntity"))
+        {
+        	this.storedTileEntityData = par1NBTTagCompound.getCompoundTag("blockTileEntity");
+        }
     }
 
     /**
@@ -218,5 +241,9 @@
         par1NBTTagCompound.setInteger("facing", this.storedOrientation);
         par1NBTTagCompound.setFloat("progress", this.lastProgress);
         par1NBTTagCompound.setBoolean("extending", this.extending);
+        if (this.storedTileEntityData != null)
+        {
+        	par1NBTTagCompound.setCompoundTag("blockTileEntity", this.storedTileEntityData);
+        }
     }
 }
